Add stricter regex for validating FR, DE, RO postal codes

johnnyxlemonade opened the pull request:
Updated the postal code validation regex to enforce the specific structure of French postal codes. The new regex ensures that the first two digits fall within the valid ranges (01–95 for metropolitan areas and 96–98 for overseas territories), while still allowing any valid 5-digit code. This improves accuracy over the previous general 5-digit validation, which allowed invalid codes.

johnnyxlemonade commented:
Updated the postal code validation logic to enforce the specific structure of German postal codes using a more precise regular expression. This regex ensures that only valid postal codes are accepted, adhering to the defined formats for various regions in Germany. This change improves data integrity and prevents invalid postal code entries.

johnnyxlemonade commented:
###  Update Romanian postal code validation with regional checks

In this PR, the validation for Romanian postal codes has been improved. Previously, the validation used a simple regex (`/^[0-9]{6}$/`) that only ensured the postal code was exactly six digits long. However, this approach did not take into account the actual regional structure of postal codes in Romania.

#### Changes made:
- **New Regex:** The updated regex now checks that the first two digits of the postal code correspond to valid regional codes (01-92). This ensures that only postal codes from recognized Romanian regions are accepted.
- **Why:** This change enhances the accuracy of postal code validation, preventing the acceptance of invalid codes that do not match real Romanian locations.

The new validation not only checks the format but also enforces a higher level of data integrity by matching against actual regional codes. This update is important for any application that handles Romanian addresses or postal information.

rubentebogt commented:
Looks cool, i'm not the maintainer of this repo so i have no say in this but i would maybe also add the new format's and regional codes to the corresponding tests in the test files. Also those that should not pass anymore now. 

Also maybe check some added newlines that are not in the other format classes. Looking forward to seeing this released 👍

BenMorel commented:
Hi @johnnyxlemonade, thanks for your PR!

I agree with @rubentebogt, we need to update tests (`XXFormatterTest`, where `XX` is the country code) to ensure that, at the very least, formats that were erroneously accepted before, are now rejected.

Can you please add those?

BenMorel commented on src/Formatter/FRFormatter.php:
```diff
@@ -11,13 +11,20 @@
     {
-        if (preg_match('/^[0-9]{5}$/', $postcode) !== 1) {
+
+        if (preg_match('/^(?:0[1-9]|[1-8]\d|9[0-8])\d{3}$/', $postcode) !== 1) {
```
Please avoid mixing `\d` with ranges, I think it hurts readability. Let's keep `[0-9]` everywhere.

johnnyxlemonade replied:
I agree with this perspective, as even if the given numbers are not in a valid range, using [0-9] enhances clarity.

johnnyxlemonade commented:
> Hi @johnnyxlemonade, thanks for your PR!
> 
> I agree with @rubentebogt, we need to update tests (`XXFormatterTest`, where `XX` is the country code) to ensure that, at the very least, formats that were erroneously accepted before, are now rejected.
> 
> Can you please add those?


Hi BenMorel,

Thank you for your message! I’ve added and updated the tests in XXFormatterTest to ensure that formats that were erroneously accepted before are now rejected. I apologize for the complications during the testing process; there were quite a few unnecessary commits because I overlooked a condition in the regular expression.

Thanks for your understanding!

BenMorel commented on tests/Formatter/DEFormatterTest.php:
```diff
@@ -36,6 +36,51 @@ public function providerFormat() : array
+            ['010101', null],
+            ['000000', null],
+            ['(123456)', null],
+            ['!23456', null],
```
The data passed to `CountryPostcodeFormatter::format()` can only contain uppercase alphanumeric characters, so testing with other characters is redundant here.

(we may revisit this in #17 though!)

BenMorel commented on src/Formatter/FRFormatter.php:
```diff
@@ -11,12 +11,19 @@
     public function format(string $postcode) : ?string
     {
+
         if (preg_match('/^[0-9]{5}$/', $postcode) !== 1) {
```
Have you removed the updated regex for `FR` by mistake?

BenMorel replied:
Also, nitpick: please try to keep formatting intact (not adding newlines), we don't have a CS fixer on this project yet, so let's try to keep it tidy manually!
